cpp/pipelines/add_subsystem_tags.cc

```diff
@@ -425,27 +424,100 @@ void AddSubsystemTags(MARC::Reader * const marc_reader, MARC::Writer * const mar
+    unsigned record_count(0);
+    while (const MARC::Record record = marc_reader->read()) {
+        ++record_count;
+        for (auto tag_to_check : tags_to_check) {
+            for (auto &field : record.getTagRange(tag_to_check)) {
+                const auto subfields(field.getSubfields());
+                for (const auto &subfield : subfields) {
+                    if (subfield.code_ == '0') {
+                        const std::string author(subfield.value_);
+                        if (likely(not author.empty())) {
+                            if (not StringUtil::StartsWith(author, "(DE-627)"))
+                                continue;
+                            std::string author_id(author.substr(std::string("(DE-627)").length()));
+                            auto it_author = authors->find(author_id);
+                            std::set<std::string> instances;
+                            if (it_author != authors->end())
+                                instances = it_author->second;
+                            
+                            if (IsRelBibRecord(record))
+                                instances.emplace("r");
+                            if (IsCanonLawRecord(record, canon_law_gnd_numbers))
+                                instances.emplace("c");
+                            if (IsBibleStudiesRecord(record, bible_studies_gnd_numbers))
+                                instances.emplace("b");
+                            
+                            if (authors->find(author_id) == authors->end())
+                                authors->emplace(author_id, instances);
+                            else
+                                (*authors)[author_id] = instances;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+void TagAuthors(MARC::Reader * const authority_reader, MARC::Writer * const authority_writer, std::map<std::string, std::set<std::string>> &authors) {
+    while (MARC::Record record = authority_reader->read()) {
+         auto it_authors = authors.find(record.getControlNumber());
+         if (it_authors != authors.end()) {
+             std::vector<MARC::Subfield> tits{ { 'a', "ixtheo" } };
+             std::set<std::string> instances = it_authors->second;
+             if (instances.find("r") != instances.end())
+                tits.push_back({ 'a', "relbib" });
+             if (instances.find("b") != instances.end())
+                tits.push_back({ 'a', "biblestudies" });
+             if (instances.find("c") != instances.end())
+                tits.push_back({ 'a', "canonlaw" });
+            record.insertField("TIT", tits);
+         }
+         authority_writer->write(record);
+    }
+}
+
+
 } //unnamed namespace
 
 
 int Main(int argc, char **argv) {
-    if (argc != 4)
-        ::Usage("marc_input authority_records marc_output");
+    if (argc != 5)
+        ::Usage("marc_input authority_records marc_output authority_output");
```

Comment:
maybe we should rename "authority_records" to "authority_input" to keep things consistent